### ISSUE TITLE
Enhance the specification to detail how exceptions are to be thrown

### ISSUE THREAD

**Issue by smoyer64** [eclipse/ConfigJSR#87]
Currently the specification only references how to handle errors as a side-effect of the desired behavior with the following two behaviors identified:

1. If a required configuration parameter is missing, a ``DeploymentException`` should be thrown.

2. If an implicit converter cannot convert a value, an ``IllegalArgumentException`` should be thrown (really this applys to explicit converters too since the ``Converter`` interface includes the equivalent throws statement).

I'd like to suggest that this leaves a lot of latitude in how these ``Exceptions`` are thrown and specifically how the configuration's deficits are passed back to the developer (and ultimately to the user).  Ultimately, the entire state of the configuration system should be evaluated and all the deficits enumerated in a way that makes it easy to iterate over the problems.  I think the ``javax.validation`` system does a pretty good job as it uses a ``ConstraintViolationException`` to pass along a ``Set<ConstrainViolation<?>>`` that can be easily parsed.  I'd suggest that each validation problem, whether it's a missing required parameter or an issue with a converter should contain a field that only contains the parameter's name as well as a message describing the problem.

As a concrete example, the Wildfly Swarm implementation of the ``microprofile-config`` API produces a ``DeploymentException`` with the following message:

```shell
org.jboss.weld.exceptions.DeploymentException: Error while validating Configuration
No Config Value exists for ldap.service.host
No Config Value exists for ldap.client.password
No Config Value exists for ldap.client.dn
```

This message actually has two problems as follows:

1. If this message is subsequently printed out to a logging system, many static code analysis tools will flag this as a security issue.

2. If the developer wants to catch this message and provide additional information and/or formatting to the message, parsing these lines out of the message requires extra code.

Finally, since JSR-382 is intended to encompass both Java SE and Java EE (yeah I know) projects, it's very likely that the developer will want to provide a customized help message for a CLI project or a detailed error dialog for a JavaFX (AWT, Swing, etc) project when a configuration is invalid in some way.  It would be great to standardize this across all implementations!

**Comment by struberg** [comment]
Hi @smoyer64 ! 

First of all a warm welcome!

I'm not sure if the analogy to BeanValidation is really that good. In BeanValidation it is _expected_ that the validation sometimes fails. 
But if a configuration value cannot be converted then it is really a setup issue of the environment or the application, isn't?. 
And this must imo lead to blowing up, isn't? So I personally don't see any necessity to transport this information in a kind of human readable form.
But of course this is just a first gut feeling. Would need to see more benefits we might get out from improving this area. You know, we always try to keep the surface of our API as minimal as possible.

Could you also probably elaborate what you mean by security issue?

**Comment by smoyer64** [comment]
As I play around with SmallRye's microprofile-config implementation, I think it's also worth specifying that __all__ validation including conversion should be completed and collected into a single ``Exception`` before it's thrown.

**Comment by struberg** [comment]
@smoyer64 that's what we do in geronimo-config: https://github.com/apache/geronimo-config/blob/ConfigJSR/impl/src/main/java/org/apache/geronimo/config/cdi/ConfigExtension.java#L114

**Comment by smoyer64** [comment]
@struberg If you run a CLI program that uses an implementation of this specification, wouldn't it be nicer if the developer could produce a human-readable error message instead of just having the application blow-up (print a stack-trace to the screen?

Here's a link to [Log Injection](https://www.owasp.org/index.php/Log_Injection) as defined by [OWASP](https://www.owasp.org).

**Comment by smoyer64** [comment]
@struberg That looks just like the SmallRye (nee Wildfly Swarm) microprofile-config code at https://github.com/smallrye/smallrye-config/blob/c9c1d5b5705c5a5a580b94b62f2b2e89b60b5633/implementation/src/main/java/io/smallrye/config/inject/ConfigExtension.java#L126

**Comment by smoyer64** [comment]
@struberg I knew that Geronimo was the original and didn't mean to imply it was copied - and with the Apache license I wasn't concerned.  It's a bit uncool to put Copyright Redhat in the TOF comments (the file I added assigned the copyright to the ASF but that may not survive).  I had __NOT__ noticed that the NOTICED provided attribution (nice!).

I love forks but it obviously works best when the forks tend to converge again.  In any case, I'm hoping you'll see some merit in the PR I've submitted to SmallRye.  Quite a few of the ``microprofile`` modules are applicable to Java SE projects that adhere to the 12-factor app guidelines so I've been pushing to make life easier for Java SE projects ... not trying to be a pain!
